axum-extra/src/routing/typed.rs

```diff
@@ -0,0 +1,171 @@
+use super::sealed::Sealed;
+
+/// A type safe path.
+///
+/// This is used to statically connect a path to its corresponding handler using
+/// [`RouterExt::typed_get`], [`RouterExt::typed_post`], etc.
+///
+/// # Example
+///
+/// ```rust
+/// use serde::Deserialize;
+/// use axum::{Router, extract::Json};
+/// use axum_extra::routing::{
+///     TypedPath,
+///     RouterExt, // for `Router::typed_*`
+/// };
+///
+/// // A type safe route with `/users/:id` as its associated path.
+/// #[derive(TypedPath, Deserialize)]
+/// #[typed_path("/users/:id")]
+/// struct UsersMember {
+///     id: u32,
+/// }
+///
+/// // A regular handler function that takes `UsersMember` as the first argument
+/// // and thus creates a typed connection between this handler and the `/users/:id` path.
+/// //
+/// // The `TypedPath` must be the first argument to the function.
+/// async fn users_show(
+///     UsersMember { id }: UsersMember,
+/// ) {
+///     // ...
+/// }
+///
+/// let app = Router::new()
+///     // Add our typed route to the router.
+///     //
+///     // The path will be inferred to `/users/:id` since `users_show`'s
+///     // first argument is `UsersMember` which implements `TypedPath`
+///     .typed_get(users_show)
+///     .typed_post(users_create)
+///     .typed_delete(users_destroy);
+///
+/// #[derive(TypedPath)]
+/// #[typed_path("/users")]
+/// struct UsersCollection;
+///
+/// #[derive(Deserialize)]
+/// struct UsersCreatePayload { /* ... */ }
+///
+/// async fn users_create(
+///     _: UsersCollection,
+///     // Our handlers can accept other extractors.
+///     Json(payload): Json<UsersCreatePayload>,
+/// ) {
+///     // ...
+/// }
+///
+/// async fn users_destroy(_: UsersCollection) { /* ... */ }
+///
+/// #
+/// # let app: Router<axum::body::Body> = app;
+/// ```
+///
+/// # Using `#[derive(TypedPath)]`
+///
+/// While `TypedPath` can be implemented manually its _highly_ recommended to derive it:
+///
+/// ```
+/// use serde::Deserialize;
+/// use axum_extra::routing::TypedPath;
+///
+/// #[derive(TypedPath, Deserialize)]
+/// #[typed_path("/users/:id")]
+/// struct UsersMember {
+///     id: u32,
+/// }
+/// ```
+///
+/// The macro expands to:
+///
+/// - A `TypedPath` implementation.
+/// - A [`Display`] implementation that interpolates the captures. This can be used to, among other
+/// things, create links to known paths and have them verified statically.
```

Comment:
I think it should be noted somewhere that the types of the fields of the struct need to implement `Display`, and that `Display` and `Deserialize` for those types need to agree on the string representation. (maybe `Serialize` should be used instead of `Display` internally?)

Reply:
Thats a good point! I'll add a note about that.

Using `Serialize` instead is interesting 🤔 I guess it would need to be serializer that only supports simple types like strings, numbers, and bools. Things that make sense in path segments and aren't nested.

Reply:
Added in feb86f6bb367f